fxa-rust-client/src/lib.rs

```diff
@@ -0,0 +1,502 @@
+    }
+
+    pub fn begin_oauth_flow(
+        &mut self,
+        redirect_uri: &str,
+        scopes: &[&str],
+        wants_keys: bool,
+    ) -> Result<String> {
+        let state = FirefoxAccount::random_base64_url_string(16);
+        let code_verifier = FirefoxAccount::random_base64_url_string(43);
+        let code_challenge = hash(MessageDigest::sha256(), &code_verifier.as_bytes())?;
+        let code_challenge = base64::encode_config(&code_challenge, base64::URL_SAFE_NO_PAD);
+        let mut url = self.state.config.content_url_path("oauth/signin")?;
+        url.query_pairs_mut()
+            .append_pair("client_id", &self.state.client_id)
+            .append_pair("redirect_uri", &redirect_uri)
+            .append_pair("scope", &scopes.join(" "))
+            .append_pair("response_type", "code")
+            .append_pair("state", &state)
+            .append_pair("code_challenge_method", "S256")
+            .append_pair("code_challenge", &code_challenge)
+            .append_pair("access_type", "offline");
+        let jwk = match wants_keys {
+            true => {
+                let jwk = JWK::from_random_ec(JWKECCurve::P256)?;
+                let jwk_json = jwk.to_json(false)?.to_string();
+                let keys_jwk = base64::encode_config(&jwk_json, base64::URL_SAFE_NO_PAD);
+                url.query_pairs_mut().append_pair("keys_jwk", &keys_jwk);
+                Some(jwk)
+            }
+            false => None,
+        };
+        let authorization_uri = url.to_string();
+        // TODO: FxA doesn't accept spaces encoded as + :(
+        let authorization_uri = authorization_uri.replace("+", "%20");
+        self.flow_store.insert(
+            state.clone(), // Since state is supposed to be unique, we use it to key our flows.
+            OAuthFlow { jwk, code_verifier },
+        );
+        Ok(authorization_uri)
+    }
+
+    pub fn complete_oauth_flow(&mut self, code: &str, state: &str) -> Result<OAuthInfo> {
+        let resp;
+        // Needs non-lexical borrow checking.
+        {
+            let flow = match self.flow_store.get(state) {
+                Some(flow) => flow,
+                None => bail!(ErrorKind::UnknownOAuthState),
+            };
+            let client = Client::new(&self.state.config);
+            resp = client.oauth_token_with_code(&code, &flow.code_verifier, &self.state.client_id)?;
+        }
+        let oauth_flow = match self.flow_store.remove(state) {
+            Some(oauth_flow) => oauth_flow,
+            None => bail!(ErrorKind::UnknownOAuthState),
+        };
+        self.handle_oauth_token_response(resp, oauth_flow.jwk)
+    }
+
+    fn handle_oauth_token_response(
+        &mut self,
+        resp: OAuthTokenResponse,
+        jwk: Option<JWK>,
+    ) -> Result<OAuthInfo> {
+        let granted_scopes = resp.scope.split(" ").map(|s| s.to_string()).collect();
+        // This assumes that if the server returns keys_jwe, the jwk argument is Some.
```

Comment:
Is it also worth erroring out if the opposite occurs, i.e. we were expecting keys but the server didn't return any?